src/mscorlib/corefx/System/Runtime/InteropServices/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Allows limited thread safe reuse of heap buffers to limit memory pressure.
+    /// 
+    /// This cache does not ensure that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
```

Comment:
> An important addition: this solution never needs to pin.

Pinning for synchronous PInvokes is close to free, it is much cheaper than overhead of safe handles that the current scheme is based on.

> ArrayPool does have a SpinLock

This is something to fix in ArrayPool so that it is no-brainer to use.